lib/mastodon/plugin.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+module Mastodon
+  class Plugin
+    NamedPath = Struct.new(:name, :path, :type)
+
+    attr_reader :actions, :paths, :assets
+
+    def initialize
+      @actions = Set.new
+      @paths   = Set.new
+      @assets  = Set.new
+    end
+
+    # defined by the plugin
+    def setup!
+      raise NotImplementedError.new, "Please define a setup! method"
+    end
+
+    private
+
+    def use_asset(path, outlet: nil)
+      path = path_prefix(path, relative: true)
+      @assets.add path
+      @paths.add NamedPath.new(outlet, path, :outlet) if outlet
+    end
+
+    def use_asset_directory(glob)
+      use_directory(glob) { |path| use_asset(path) }
+    end
+
+    def use_class(path)
```

Comment:
Maybe it would be good to have a lib folder inside plugin directory, and load everything from it when the plugin is activated. It seems unnecessary to explicitly load classes this way...